### PR TITLE
WIP: Refactor markdown

### PR DESCRIPTION
**Pull request by RadicalZephyr**
Here's our start at rewriting the markdown extension as proposed in #160.  Feedback or suggestions are welcome!

### CODE REVIEW

File: components/rendering/src/markdown.rs [hunk 1]
```diff
@@ -1,20 +1,121 @@
...
 use short_code::{SHORTCODE_RE, ShortCode, parse_shortcode, render_simple_shortcode};
 use table_of_contents::{TempHeader, Header, make_table_of_contents};
 
+struct GutenbergFlavoredMarkdownParser<'a, 'b, 'c> {
+    context: &'a Context<'a>,
+    errors: &'c mut Vec<Error>,
+    parser: Peekable<Parser<'b>>,
+}
+
+fn is_internal_link(link: &str) -> bool {
```
Comment (Keats) [inline components/rendering/src/markdown.rs:23]: this is only used once so I don't think it needs to be a function

File: components/rendering/src/markdown.rs [hunk 2]
```diff
@@ -1,20 +1,121 @@
...
+    link.starts_with("./")
+}
+
+impl<'a, 'b, 'c> GutenbergFlavoredMarkdownParser<'a, 'b, 'c> {
+    fn new(context: &'a Context<'a>, parser: Parser<'b>, errors: &'c mut Vec<Error>) -> GutenbergFlavoredMarkdownParser<'a, 'b, 'c> {
+        GutenbergFlavoredMarkdownParser {
+            context: context,
+            errors: errors,
```
Comment (Keats) [inline components/rendering/src/markdown.rs:31]: no need to repeat `context` and `errors`, ` {context, errors, parser: parser.peekable()}` will work

File: components/rendering/src/markdown.rs [hunk 3]
```diff
@@ -1,20 +1,121 @@
...
+                    match resolve_internal_link(href, self.context.permalinks) {
+                        Ok(url) => {
+                            Event::Start(Tag::Link(Owned(url), text.clone()))
+                        },
+                        Err(_) => {
+                            self.push_error(format!("Relative link {} not found.", href));
+                            self.empty_event()
+                        }
+                    }
+                } else{
```
Comment (Keats) [inline components/rendering/src/markdown.rs:57]: space after else

File: components/rendering/src/markdown.rs [hunk 4]
```diff
@@ -1,20 +1,121 @@
...
+                    Event::Start(Tag::Link(href.clone(), text.clone()))
```
Comment (Keats) [inline components/rendering/src/markdown.rs:58]: not possible to return the `link` without cloning?

File: components/rendering/src/markdown.rs [hunk 5]
```diff
@@ -1,20 +1,121 @@
...
+
+    fn next(&mut self) -> Option<Event<'b>> {
+        match self.parser.next() {
+            Some(event) => Some(self.process_event(event)),
+            None => None,
+        }
+    }
+}
+
+#[allow(unused_variables)]
```
Comment (Keats) [inline components/rendering/src/markdown.rs:90]: is that just while it's a wip?

**Review by Keats — COMMENTED** [review]
Looks like a good start!

**Comment by Keats** [comment]
Ah can you do the PR on the `next` branch by any chance?

File: components/rendering/src/markdown.rs (hunk 2 above)
Comment (RadicalZephyr) [reply components/rendering/src/markdown.rs:31]: whoa! I had no idea! that's awesome

File: components/rendering/src/markdown.rs (hunk 5 above)
Comment (RadicalZephyr) [reply components/rendering/src/markdown.rs:90]: yes, to make fewer warnings.  We'll make sure to remove these and fix warnings before finally submitting.

File: components/rendering/src/markdown.rs (hunk 1 above)
Comment (reillysiemens) [reply components/rendering/src/markdown.rs:23]: Yes, it's only used once, but I think the value here is being explicit about what the parsing is looking for. When simply using `link.starts_with("./")` I think it's not as apparent that what we're really looking for is an internal link.

File: components/rendering/src/markdown.rs (hunk 4 above)
Comment (reillysiemens) [reply components/rendering/src/markdown.rs:58]: The current implementation clones as well, so we just left it while it's still a WIP, but we'll go back at the end and clean up as many clones as possible.

**Comment by Keats** [comment]
I'm guessing this PR can be redone from scratch with the coming changes to the shortcode handling? Or at least significant changes

**Comment by RadicalZephyr** [comment]
Closing this for now, since we're going to regroup a bit on our approach.  We'll open a new PR once we've rebased against the next branch.

**Pull request closed without merging**